src/query/api/v1/httpd/handler.go

```diff
@@ -318,19 +308,91 @@ func (h *Handler) RegisterRoutes() error {
 					Tagged(remoteSource).
 					Tagged(experimentalAPIGroup),
 			)
-			h.router.HandleFunc(annotated.WriteURL,
-				wrapped(experimentalAnnotatedWriteHandler).ServeHTTP,
-			).Methods(annotated.WriteHTTPMethod)
+			addWrappedRoute(annotated.WriteURL, experimentalAnnotatedWriteHandler,
+				annotated.WriteHTTPMethod)
 		}
 	}
 
 	h.registerHealthEndpoints()
 	h.registerProfileEndpoints()
 	h.registerRoutesEndpoint()
 
+	// Register custom endpoints.
+	for _, custom := range h.customHandlers {
+		routeName := routeName(custom.Route(), custom.Methods())
+		route := h.router.Get(routeName)
+		var prevHandler http.Handler
+		if route != nil {
+			prevHandler = route.GetHandler()
+		}
+		customHandler, err := custom.Handler(nativeSourceOpts, prevHandler)
+		if err != nil {
+			return fmt.Errorf("failed to register custom handler with path %s: %w",
+				routeName, err)
+		}
+
+		if route == nil {
+			addWrappedRoute(custom.Route(), customHandler, custom.Methods()...)
+		} else {
+			route.Handler(wrapped(customHandler))
+		}
+	}
+
 	return nil
 }
 
+func addRoute(router *mux.Router, path string, handler http.Handler, methods []string) {
+	addRouteHandlerFn(router, path, handler.ServeHTTP, methods...)
+}
+
+func addRouteHandlerFn(
+	router *mux.Router,
+	path string,
+	handlerFn func(http.ResponseWriter, *http.Request),
+	methods ...string,
+) {
+	routeName := routeName(path, methods)
+	if previousRoute := router.Get(routeName); previousRoute != nil {
+		// route already exists.
+		return
+	}
+
+	route := router.
+		HandleFunc(path, handlerFn).
+		Name(routeName)
+
+	if len(methods) > 0 {
+		route.Methods(methods...)
+	}
+}
+
+func routeName(p string, methods []string) string {
+	if len(methods) > 1 {
```

Comment:
nit; should probably warn or error if no methods are passed in? Otherwise you can end up adding a handler with no visiblity into why it's not coming up